### PR TITLE
Timo updates

### PR DESCRIPTION
**Pull request by tjrantal**
Hello Richard,

I just read the clone, branch, upload and hence only branched once I'd already made my changes. Hopefully this works nevertheless.

I noticed a few bugs, and on occasion simply reverted to my for loops. I'm not familiar with streams and lambdas and couldn't always replicate the operations I'd done with for loops. On that note, I'm not a lambda convert yet...

I think we could work on creating a test case to ensure that bugs (i.e. unexpected analysis results) get picked up during building. I'm vaguely aware this is possible through some sort of testing library but haven't ever bothered to find out how to actually do that. I could contribute a scan of my leg, and we could check all of the results match results from this version as the analysis result check.

Regards,
-Timo

### CODE REVIEW

**Comment by rimadoma** [comment]
Thanks so much! And sorry again for causing the hassle. 

We could try to create automated tests with JUnit, Mockito and other such tools. Regression tests check - like you suggested - that predetermined, expected test image values don't change. One of the challenges would be to figure out how to disable calls to GUI (dialogs etc popping up), because that's not allowed on the Travis server.

Another possibility would be to include a small, zipped test image, and a script to perform a manual test. Bit like [here](https://github.com/bonej-org/BoneJ2/tree/ani-text/IntegrationTestLogs).

File: src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java [hunk 1]
```diff
@@ -140,21 +151,44 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
...
+		return b;
+	}
+	
+	protected byte[] clone(byte[] a){
+		byte[] b = new byte[a.length];
+		for (int i = 0;i<a.length;++i){
+			b[i] = a[i];
+		}
+		return b;
+	}
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java:190]: Why are the custom clone methods, for things that can be achieved by the `clone()` method of an array, `System.arraycopy()` or `Arrays.copyOf()`?

Comment (rimadoma) [reply src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java:190]: Why are these methods `protected`?

File: src/main/java/sc/fiji/pQCT/selectroi/SelectROI.java [hunk 2]
```diff
@@ -113,6 +118,18 @@ public SelectROI(final ScaledImageData dataIn,
...
+		/*
+		ImagePlus tempImage = new ImagePlus("peeledROI");
+		tempImage.setProcessor(new FloatProcessor(width,height,scaledImage));
+		tempImage.show();
+		
+		ImagePlus tempImage2 = new ImagePlus("sieve");
+		tempImage2.setProcessor(new ByteProcessor(width,height,sieve));
+		tempImage2.setDisplayRange(0,1);
+		tempImage2.show();
+		*/
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/selectroi/SelectROI.java:131]: Please remove any commented out code. It makes the code harder to read.

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java [hunk 3]
```diff
@@ -168,6 +217,9 @@ private void calculateRadii() {
 			// division
 			final double analysisThickness = BMD_temp.size();
 			if (analysisThickness < divisions) {
+				IJ.log("Too thin "+et
+						+" r2[et] "+r2[et]
+						+" r[et] "+r[et]);
 				break;
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java]: Before merging the PR, I'll format the code (tabs, indent etc.)

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java [hunk 4]
```diff
@@ -168,7 +164,7 @@ public DistributionAnalysis(final SelectROI roi,
 
 	// TODO Add a boolean parameter preventPeeling, and combine method with
 	// calculateRadiiNoPeeling
-	private void calculateRadii() {
+	private void calculateRadii(boolean preventPeeling) {
 		// Calculate radii in polar coordinate system originating from bone marrow
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java:218]: Parameters, local variables, fields etc. should have the `final` keyword when applicable.

File: src/main/java/sc/fiji/pQCT/analysis/DetermineAlpha.java [hunk 5]
```diff
@@ -279,7 +279,7 @@ private void determineMomentAlpha(final double[] csmiValues) {
 	private Vector<Integer> rotateIndex(final int rotationAngle) {
 		final int initialIndex = 360 - Math.abs(rotationAngle);
 		final Vector<Integer> rotateIndexVector = Stream.iterate(initialIndex,
-			i -> i++ % 360).limit(360).collect(toCollection(Vector::new));
+			i -> i+1).limit(360).map(ii -> {return ii % 360;}).collect(toCollection(Vector::new));
 		/*Flip rotateIndexVector, for e.g. comparing left to right*/
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/DetermineAlpha.java]: Can be written simply as `map(ii -> ii % 360)`

File: src/main/java/sc/fiji/pQCT/selectroi/SelectROI.java [hunk 6]
```diff
@@ -40,9 +40,9 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 import sc.fiji.pQCT.io.ScaledImageData;
 
 //Debugging
-import ij.IJ;		//Float Images
-import ij.process.FloatProcessor;		//Float Images
-import ij.process.ByteProcessor;
+//import ij.IJ;		//Float Images
+//import ij.process.FloatProcessor;		//Float Images
+//import ij.process.ByteProcessor;
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/selectroi/SelectROI.java]: Please delete unnecessary imports rather than commenting them out

File: src/main/java/sc/fiji/pQCT/utils/ResultsWriter.java [hunk 7]
```diff
@@ -171,21 +171,23 @@ public void writeHeader(final TextPanel textPanel,
...
-				headings.append("\t").append(rowStart).append(
-					" ?pericortical vBMD [mg/cm3]");
+			}
+			
+			
+			final String[] varNames = {" endocortical radius [mm]"," pericortical radius [mm]"," endocortical vBMD [mg/cm3]"," midcortical vBMD [mg/cm3]"," pericortical vBMD [mg/cm3]"};
+			
+			
+			//add variables with similar naming convention in two loops
+			for (int v = 0;v<varNames.length;++v){
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/utils/ResultsWriter.java:185]: Good change. As a tip, often a comment like this is a sign that the piece of code is a good candidate for an aptly named little method.

**Comment by rimadoma** [comment]
For future reference, have a look at this [guide[(https://github.com/erlang/otp/wiki/writing-good-commit-messages) on writing good commit messages.

**Comment by rimadoma** [comment]
Also, it's good to know that commits like

`aaaaaaa Change to myClass part1`
`bbbbbb Change to myClass part2`

Can be later combined with a `git rebase` command. The same command can be used to change commit messages after pushing (if you notice an error before pushing, you can use `git commit --amend`)

File: src/main/java/sc/fiji/pQCT/analysis/ConcentricRingAnalysis.java [hunk 8]
```diff
@@ -150,10 +152,10 @@ private void calculateRadii() {
 
 	private void rotateResults() {
 		//Calculate pericortical radii
-		double[] pRad = DistributionAnalysis.clone(rU);
+		//Double[] pRad = DistributionAnalysis.clone(rU);
 		//IJ.log(String.format("rU length %d",rU.length));
 		//IJ.log(String.format("pRad length %d",pRad.length));
-		Arrays.stream(pRad).forEach(r -> r *= pixelSpacing);
+		double[] pRad =(double []) stream(rU).map(r -> {return r *= pixelSpacing;}).toArray();
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/ConcentricRingAnalysis.java]: `map` doesn't need the return

Why the cast to `double []`?

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java [hunk 9]
```diff
@@ -242,42 +254,39 @@ private void calculateRadii(boolean preventPeeling) {
...
+			for (int j = 1; j < width-1; j++) {
 				final int index = i * width + j;
-				if (data[index] > minimum) {
-					if (i > 0 && data[(i - 1) * width + j] == minimum || j > 0 &&
-						data[(i) * width + j - 1] == minimum || i + 1 < height && data[(i +
-							1) * width + j] == minimum || j + 1 < width && data[(i) * width +
-								j + 1] == minimum)
+				if (data[index] > bgVal) {
+					if (data[(i - 1) * width + j] == bgVal | 
+						data[(i) * width + j - 1] == bgVal |
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java:313]: Potential bug here: a `|` operator does a very different thing than a `||`!

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java [hunk 10]
```diff
@@ -327,14 +336,10 @@ private double expandRadiusMulti(final double[] roi, final double threshold,
...
-		final double[] eRad = clone(rS);//rS.clone();
-		stream(eRad).forEach(r -> r *= pixelSpacing);
-		final double[] pPRad = clone(r);//r.clone();
-		stream(pPRad).forEach(r -> r *= pixelSpacing);
-		final double[] pERad = clone(r2);//r2.clone();
-		stream(pERad).forEach(r -> r *= pixelSpacing);
+		final double[] pRad = (double []) stream(rU).map(r -> {return r *= pixelSpacing;}).toArray();
+		final double[] eRad =  (double []) stream(rS).map(r -> {return r *= pixelSpacing;}).toArray();
+		final double[] pPRad = (double []) stream(r).map(r -> {return r *= pixelSpacing;}).toArray();
+		final double[] pERad = (double []) stream(r2).map(r -> {return r *= pixelSpacing;}).toArray();
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java]: No returns or casts needed (see above).

File: src/main/java/sc/fiji/pQCT/selectroi/SelectROI.java [hunk 11]
```diff
@@ -136,20 +134,22 @@ public SelectROI(final ScaledImageData dataIn,
...
-				}
-				if (scaledImage[index] >= areaThreshold) {
-					cortexAreaRoiI.add(i);
-					cortexAreaRoiJ.add(j);
-				}
-				if (scaledImage[index] >= BMDthreshold) {
-					cortexROI[index] = scaledImage[index];
-					cortexRoiI.add(i);
-					cortexRoiJ.add(j);
+				}else{
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/selectroi/SelectROI.java]: Why the change in `if` blocks? Nothing after a `continue` gets executed, so the `else` is not needed.

File: src/main/java/sc/fiji/pQCT/analysis/SoftTissueAnalysis.java [hunk 12]
```diff
@@ -65,6 +68,13 @@ public SoftTissueAnalysis(final SelectSoftROI roi) {
 		final double[][] stats = IntStream.range(1, 8).mapToObj(
 			typeStatistics::apply).map(summary -> new double[] { summary.getCount(),
 				summary.getAverage() }).toArray(double[][]::new);
+				
+		for (int i = 0;i<stats.length;++i){
+			IJ.log(String.format("Stats %d one %.2f two %.2f",i,stats[i][0],stats[i][1]));
+		}
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/SoftTissueAnalysis.java]: Left from debugging? Is it necessary for the user to see this?

File: src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java [hunk 13]
```diff
@@ -385,7 +385,7 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 		return returnVectorVectorPointer;
 	}
 
-	private byte[] dilate(final byte[] data, final byte dilateVal, final byte min,
+	protected byte[] dilate(final byte[] data, final byte dilateVal, final byte min,
 		final byte temp)
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java:389]: Why `protected`?

File: src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java [hunk 14]
```diff
@@ -173,15 +173,15 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
...
 		double[] b = new double[a.length];
 		for (int i = 0;i<a.length;++i){
 			b[i] = a[i];
 		}
 		return b;
 	}
 	
-	protected byte[] clone(byte[] a){
+	protected static byte[] clone(byte[] a){
 		byte[] b = new byte[a.length];
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java]: Unnecessary custom `clone()` methods.

File: src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java [hunk 15]
```diff
@@ -1084,16 +1084,18 @@ else if (result[i + j * width] != 1) {
 		return sleeve;
 	}
 
-	void erode(final byte[] data) {
+	public byte[] erode(byte[] data) {
 		// Erode algorithm
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java:1088]: Why `public`? Does `erode` need to be called outside the pQCT package?

File: src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java [hunk 16]
```diff
@@ -1084,16 +1084,18 @@ else if (result[i + j * width] != 1) {
...
+			for (int j = 1; j < width-1; j++) {
 				if (data[i * width + j] > 0) {
-					if (i > 0 && data[(i - 1) * width + j] == 0 || j > 0 && data[(i) *
-						width + j - 1] == 0 || i + 1 < height && data[(i + 1) * width +
-							j] == 0 || j + 1 < width && data[(i) * width + j + 1] == 0)
+					if (	data[(i - 1)	* width + j] == 0
+						| 	data[(i) 		* width + j - 1] == 0
+						| 	data[(i + 1)	* width +j] == 0
+						| 	data[(i)		* width + j + 1] == 0
+						)
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java:1098]: Potential bug: `|` != `||`

File: src/main/java/sc/fiji/pQCT/selectroi/SelectSoftROI.java [hunk 17]
```diff
@@ -44,9 +44,14 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
...
-public class SelectSoftROI extends RoiSelector {
+//Debugging
+import ij.IJ;		//Float Images
+import ij.process.FloatProcessor;		//Float Images
+import ij.process.ByteProcessor;
 
-	private static final int RADIAL_DIVISIONS = 720;
+public class SelectSoftROI extends RoiSelector {
+	
+	private final int RADIAL_DIVISIONS = 720;
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/selectroi/SelectSoftROI.java:54]: Why can't this be static?

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java [hunk 18]
```diff
@@ -130,9 +134,22 @@ public DistributionAnalysis(final SelectROI roi,
 		marrowCenter[0] /= marrowI.size();
 		marrowCenter[1] /= marrowJ.size();
 		
-		peeledBMD = range(0, peeledSize).filter(i -> peeledROI[i] >= threshold).mapToDouble(ii -> {return peeledROI[ii];})
-			.average().orElse(0.0);
-		IJ.log("After peeling PeeledBMD "+peeledBMD);
+		peeledBMD = range(0, peeledSize).filter(i -> peeledROI[i] >= threshold).mapToDouble(ii -> {return peeledROI[ii];}).average().orElse(0.0);
+
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java]: `return` not needed in `map`

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java [hunk 19]
```diff
@@ -128,28 +121,13 @@ public DistributionAnalysis(final SelectROI roi,
...
 		for (int i = 0; i < marrowI.size(); i++) {
-			marrowCenter[0] += marrowI.get(i);
-			marrowCenter[1] += marrowJ.get(i);
+			marrowCenter[0] += (double) marrowI.get(i);
+			marrowCenter[1] += (double) marrowJ.get(i);
 		}
-		marrowCenter[0] /= marrowI.size();
-		marrowCenter[1] /= marrowJ.size();
+		marrowCenter[0] /= (double) marrowI.size();
+		marrowCenter[1] /= (double) marrowJ.size();
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java]: When assigned to a `double` array, the variables are cast implicitly

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java [hunk 20]
```diff
@@ -147,10 +147,10 @@ public DistributionAnalysis(final SelectROI roi,
...
 		maxRadius = range(0, peeledSize).filter(i -> peeledROI[i] >= threshold)
 			.mapToDouble(index -> {
-				final int j = index / width;
+				final int j = (int) Math.floor(index / width);
 				final int i = index - width * j;
-				final double x = i - marrowCenter[0];
-				final double y = j - marrowCenter[1];
+				final double x = ((double) i) - marrowCenter[0];
+				final double y = ((double) j) - marrowCenter[1];
 				//IJ.log("X "+x+" Y "+y+" mx "+x+" my "+y+" rad "+Math.sqrt(x * x + y * y));
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java]: Doubles implicitly cast in assignment, `(double)` not needed.

Comment (rimadoma) [reply src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java]: Also, if any variable in a calculation is double, all the other variables are implicitly cast to double.

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java [hunk 21]
```diff
@@ -126,6 +126,7 @@ public DistributionAnalysis(final SelectROI roi,
 		}
 		marrowCenter[0] /= (double) marrowI.size();
 		marrowCenter[1] /= (double) marrowJ.size();
+		IJ.log("C0 "+marrowCenter[0]+" C1 "+marrowCenter[1]);
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java]: Left in code from debugging? Remove if not necessary for the user to see.

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java (hunk 20 above)
Comment (rimadoma) [reply src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java]: Why are variables not `final` anymore?

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java [hunk 22]
```diff
@@ -340,13 +381,28 @@ private void rotateResults() {
...
 				// Cortex
 				endoCorticalBMDs[pp] += bMDJ.get(0)[index] / sectorWidth;
 				midCorticalBMDs[pp] += bMDJ.get(1)[index] / sectorWidth;
 				periCorticalBMDs[pp] += bMDJ.get(2)[index] / sectorWidth;
+				
+				IJ.log("NEW pp "+pp+" sector dd "+dd+" pind "+index +" eRad "+endocorticalRadii[pp]+" pRad "+ pericorticalRadii[pp]
+				+" eBMD "+ endoCorticalBMDs[pp]
+				+" mBMD "+ midCorticalBMDs[pp]
+				+" pBMD "+ periCorticalBMDs[pp]);
 				//IJ.log("pp "+pp+" index "+index);
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java]: Left from debugging?

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java [hunk 23]
```diff
@@ -276,7 +278,7 @@ private void calculateRadii(boolean preventPeeling) {
 			}
 			
 			//Get the BMDs here
-
+			IJ.log(String.format("theta %d %.2f endo %.2f peri %.2f",et,theta[et],r2[et],r[et]));
 			// Dividing the cortex to three divisions -> save the mean vBMD for each
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java]: Debugging?

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java [hunk 24]
```diff
@@ -220,24 +220,40 @@ private void calculateRadii(boolean preventPeeling) {
...
+		final double y = marrowCenter[1];
+		
 		for (int et = 0; et < 360; ++et) {
 			final Vector<Double> BMD_temp = new Vector<>();
 			theta[et] = Math.PI / 180.0 *((double) et);
+			
 			if (et > 0) {
-				r[et] = rS[et - 1] / 2.0;
+				r[et] = Math.round((rS[et - 1] / 2.0)*10d)/10d;
 			}
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java]: 10d can be replaced with 10.0. Decimal literals are always `double` by default.

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java [hunk 25]
```diff
@@ -195,6 +195,7 @@ public DistributionAnalysis(final SelectROI roi,
 				return Math.sqrt(x * x + y * y);
 			}).max().orElse(0.0);
 			
+		maxRadius = Math.round(maxRadius*10d)/10d;	//Needs to be rounded to 0.1
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java]: 10d can be just floating point literals

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java [hunk 26]
```diff
@@ -299,6 +295,7 @@ private void calculateRadii(boolean preventPeeling) {
 				bMDJ.get(div)[et] /= (double) mo;
 				
 			}
+			IJ.log(String.format(Locale.ROOT,"theta %d %.2f endo %.2f peri %.2f eBMD %.2f mBMD %.2f pBMD %.2f",et,theta[et],r2[et],r[et],bMDJ.get(0)[et],bMDJ.get(1)[et],bMDJ.get(2)[et]));
 		}
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java]: Debugging?

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java [hunk 27]
```diff
@@ -83,7 +83,7 @@ public DistributionAnalysis(final SelectROI roi,
 		//}
 		
 		sectorWidth = details.sectorWidth;
-		final int size = (int) (360 / sectorWidth);
+		final int size = (int) (360d / sectorWidth);
 		endocorticalRadii = new double[size];
```
Comment (rimadoma) [inline src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java]: Can be just a floating point literal

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java (hunk 4 above)
Comment (rimadoma) [reply src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java:218]: `final` communicates the intent that the variables are read-only. Also it can prevent bugs where you accidentally change the value of a key variable.

File: src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java (hunk 1 above)
Comment (tjrantal) [reply src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java:190]: I added a custom clone method because the .clone() did not actually produce a copy of the array but rather returned (in C-language terms) a pointer to the original array. When I then modified the contents of the copy, it also modified the original array I wanted not to modify. I think I've stumbled upon similar problems with Arrays.copyOf() in the past, and decided the only way I'll trust the code is by manually copying the contents to a new array. Having said that, System.arraycopy may establish the effect I wanted to have but I didn't remember that method when modfying the code.

Re protected, no particular reason, I've heard of the idea of trying to limit access if possible so used protected rather than public. I am not a programmer so I am inconsistent with these sorts of things...

File: src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java (hunk 13 above)
Comment (tjrantal) [reply src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java:389]: I could not call the private method from a class that extended this class.

File: src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java (hunk 15 above)
Comment (tjrantal) [reply src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java:1088]: No. I think I thought of possibly calling erode from another class but don't think I ever have.

**Comment by tjrantal** [comment]
Did you want me to take out the commented out code bits, and the other things you'd highlighted (protected, final etc.)?

Re: We could try to create automated tests...

Sounds good. I'll look into  this within the next few months.

File: src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java (hunk 1 above)
Comment (rimadoma) [reply src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java:190]: This is true for arrays of objects - they create what are called shallow copies. You get a new array, but the object references i.e. pointers point to the same place as the elements in the original array. However, with an array of primitives such as byte[] or double[] (NB not the case for Double[]!) you do get a true copy. See: https://stackoverflow.com/questions/5821851/does-calling-clone-on-an-array-also-clone-its-contents

File: src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java (hunk 13 above)
Comment (rimadoma) [reply src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java:389]: Fair enough. I would be cautious with using `protected` visibility. It's hard to keep track of shared fields and methods between super and sub classes.

File: src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java (hunk 15 above)
Comment (rimadoma) [reply src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java:1088]: Fair enough. Again it's best to be cautious and only make things `public` when you're sure you need access to them.

**Comment by rimadoma** [comment]
I can do the `final` and formatting bits, because I've set my IDE so that it's just a couple of clicks. If you wouldn't mind changing those visibility modifiers, and explaining why you're using `|` instead of `||` please?
Also, if you have the time, could you write a short summary to the PR message to detail what issues these commits fix?

File: src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java (hunk 1 above)
Comment (rimadoma) [reply src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java:190]: `protected` is weaker than `private`, and it's actually quite rare that you really need it instead of private.

**Comment by rimadoma** [comment]
Hi again Timo, I made those changes we discussed [here](https://github.com/fiji/pQCT/tree/tjrantal-timo-updates) (not quite sure why the commits don't show up here, but I'm still learning how to manage PRs from forks). If we could clarify those if statements with `|` instead of `||` I'd be happy to merge, release and upload the patch to the update site.

**Comment by tjrantal** [comment]
I'm a bit snowed under at the moment. I'll try to get to this asap. I changed | to || because I didn't think it would make a difference in the use case. Essentially the bit value (i.e. the bitwise or) should evaluate to 0 if none of the conditions are true, and hence it should result in it not mattering. As you mentioned this is a conspicuous potential bug (I often make mistakes when combining multiple logical conditions...) so I will have to review to confirm whether or not these are as intended. I'll pull from your branch, and make a new pull-request (I don't think it'll update this one) once I get to it.

**Comment by rimadoma** [comment]
Well, if that's the only thing that needs to change, maybe just give me your A-OK when you've made sure everything looks OK? I'll then do the fixes left and merge the code.

It's true that logically `|` v. `||` probably won't make a difference here, but imo the former is rather obscure (especially for people with no background in C).

P.S. Literally or figuratively snowed under? Jyvaskyla still looks quite cold!

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java (hunk 9 above)
Comment (tjrantal) [reply src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java:313]: In this particular case either should result in the same result. This if should evaluate to true if any of the ... == bgVal is true. If || is faster I'd suggest that otherwise I'd leave as is. I didn't have a good reason to to change this to bitwise |, I think it just fitted on one row on my display when I took the second | off... and made it easier to inspect.

File: src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java (hunk 16 above)
Comment (tjrantal) [reply src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java:1098]: This is akin to the previous similar comment. Again I want the if to evaluate to true if any of the ... == 0 is true -> should be the same result independent of whether it is | or ||.

**Comment by tjrantal** [comment]
I responded to the comments above. I'm more than happy with || for easier readability as you outlined, and as far as I can tell it will work with either (it was indeed || previously, I swapped it to | but I'm not entirely sure why exactly).

re snowed under, it is indeed a bit nippy overnight :)

File: src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java (hunk 9 above)
Comment (imagejan) [reply src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java:313]: > This if should evaluate to true if any of the ... == bgVal is true.

With the important difference that `||` will not test the follow-up statements as soon as one of them evaluates to `true`. See also [this SO answer](https://stackoverflow.com/a/1808879/1919049).

Comment (ctrueden) [reply src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java:313]: Always use `||` for boolean logic. We actually hit a bug in Java a few years back, where some ImgLib code used `|` with booleans by accident, and it caused the Java runtime to segfault. IMHO, the java compiler should not even allow `|` with boolean operands—there is no advantage except for the non-short-circuiting @imagejan mentioned... but if you are relying on side effects in your conditional statements, the code is probably too opaque anyway.

Comment (tjrantal) [reply src/main/java/sc/fiji/pQCT/analysis/DistributionAnalysis.java:313]: I was familiar with the distinction between || and | but it never occurred to me that it might not be a good idea to do a bitwise or on booleans. I just assumed false is 0x00 and true 0x01. Should not assume... On that point, I've had Java segfaults I could not figure out (occured when running with Linux but not on Windows machine) in other code I've written -> will need to check if this is the culprit. Live and learn, thank you for the comments!
